[FLINK-3921] StringParser encoding

rekhajoshm opened the pull request:
Corrected StringParser encoding

greghogan commented:
The only internal usage of `StringParser` is from `GenericCsvInputFormat`. Should we make the encoding configurable in `GenericCsvInputFormat` with a default of US-ASCII? This could then be overridden in an additional constructor of `StringParser`.

Should the same changes be made to `StringValueParser`?

@rekhajoshm @fhueske @StephanEwen :+1: :-1:?

rekhajoshm commented:
Thanks @greghogan for your inputs. updated. Please have a look. 
Please note though the configurable Charset will have no impact on StringValueParser(or other parsers) except StringParser.In SVParser we explicitly call setValueAscii(), and hence works with ascii only. @StephanEwen @fhueske

StephanEwen commented:
Let's make this configurable on `GenericCsvInputFormat`, with the default of "UTF-8" (this is what we use in other places as the default, too).

The charset affects the
  - delimiter
  - comments
  - parsers

I think the `FieldParser` should have a `setCharset()` method. That way, we need not pass the charset to every method call.

This would also need a test.

rekhajoshm commented:
Sorry for delay, got busy. 
Thank you @StephanEwen for your review.updated.Please have a look.thanks cc @greghogan

greghogan commented on flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java:
```diff
@@ -106,6 +106,11 @@ protected GenericCsvInputFormat() {
+
+	protected GenericCsvInputFormat(Path filePath, Charset charset) {
+		super(filePath);
+		this.charset = charset != null ? charset : Charset.forName("UTF-8");
```
Would this be better as `this.charset = Preconditions.checkNotNull(charset);` since the user should use `GenericCsvInputFormat(Path)` when using the default charset?

greghogan commented:
`StringParser` and `StringValueParser` have the method `enableQuotedStringParsing` which is called by `GenericCsvInputFormat.open` only for these types. We should consider doing the same for `setCharset` rather than adding it to `FieldParser`.

greghogan commented:
Hi @rekhajoshm have you had a chance to look at the last two comments?

rekhajoshm commented:
Hi @greghogan sorry, lost thread on this, updated for Preconditions check now. Unless i am missing something, the enableQuotedStringParsing() is present in subclasses as it is slightly different than the parent and involves if-else-casting to be invoked.But in case of setCharset it is same for all possible subclasses and does not need to be.Hope you agree.Thanks.

greghogan commented on flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java:
```diff
@@ -314,6 +320,25 @@ protected void setFieldsGeneric(boolean[] includedMask, Class<?>[] fieldTypes) {
 	}
 
+	/**
+	 * Gets the Charset for the parser.Default is set to UTF-8
```
Should we spell out `charset` as "character set" in the documentation? Also, space after period and missing period at end.

greghogan commented on flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java:
```diff
@@ -314,6 +320,25 @@ protected void setFieldsGeneric(boolean[] includedMask, Class<?>[] fieldTypes) {
+	 * Sets the charset of the parser. Called by subclasses of the parser to set the type of charset
+	 * when doing a parse.
+	 *
+	 * @param charset The charset  to set.
```
Double space.

greghogan commented on flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java:
```diff
@@ -314,6 +320,25 @@ protected void setFieldsGeneric(boolean[] includedMask, Class<?>[] fieldTypes) {
+	 * @param charset The charset  to set.
+	 */
+	protected void setCharset(Charset charset){
+		this.charset = charset != null ? charset : Charset.forName("UTF-8");
```
Use `Preconditions.checkNotNull`?

greghogan commented on flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java:
```diff
@@ -314,6 +320,25 @@ protected void setFieldsGeneric(boolean[] includedMask, Class<?>[] fieldTypes) {
+	 *
+	 * @param charset The charset  to set.
+	 */
+	protected void setCharset(Charset charset){
```
Add space between `)` and `{` (and I count two more instances).

greghogan commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -75,8 +76,30 @@
 	}
-	
+
+	private Charset charset = Charset.forName("US-ASCII");
```
Should this also default to `UFT-8`?

greghogan commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -75,8 +76,30 @@
+	 * @return The index of the next delimiter, if the field was parsed correctly. A value less than 0 otherwise.
+	 */
+	public int parseField(byte[] bytes, int startPos, int limit, byte[] delim, T reuse, Charset charset){
+		this.charset = charset;
```
Is this method needed? `GenericCsvInputFormat.open` can `setCharset` on each newly instantiated `FieldParser`, and in the case where a user decided to change charset on an open file `GenericCsvInputFormat.setCharset` could go through and `setCharset` on the list of `FieldParser`s.

rekhajoshm commented:
Thanks @greghogan for your review. Please have a look. thanks!

greghogan commented:
Apologies for the long delay. I'd like to attempt to summarize this ticket and pull request to validate my understanding.

Previously StringParser was using the system encoding and `GenericCsvInputFormat` was using UTF-8 for the delimiter and an overloadable UTF-8 for the comment prefix.

StringParser's quoteCharacter remains a `byte` with no encoding.

Now GenericCsvInputFormat can be configured with a charset which is used for the delimiter, comment prefix, and field parsers (only used in StringParser).

Should `setCommentPrefix(String commentPrefix, Charset charset)` and `setCommentPrefix(String commentPrefix, String charsetName)` be removed from `GenericCsvInputFormat`? Would different encodings be used on the same file?

Allow the user to set the character encoding in `CsvReader` which would be applied in `CsvReader.configureInputFormat`?

Are the new tests checking the encoding? The test strings are using using characters common to UTF-8 and ASCII. We could instead use one of the UTF-16 encodings from https://docs.oracle.com/javase/7/docs/api/java/nio/charset/Charset.html

greghogan commented:
@rekhajoshm have you had a chance to look again at this PR? The 1.2 feature freeze will be in a few weeks.

rekhajoshm commented:
Agree @greghogan , we started this just for StringParser encoding :-)
I am in tight schedule, however did a quick update for comments, please have a look.thank you.

- Removed unused methods on setCommentPrefix
- Added charset functionality for user to set charset encoding on CsvReader
- Updated test

greghogan commented:
I am working on merging this ..
